[ASan][libc++] Update string ASan annotations to zero-overhead

AdvenamTacet opened the pull request:
This commit introduces zero-overhead string ASan annotations by wrapping every call to ASan helper functions inside a constexpr if statement. This ensures that ASan checks are only performed at runtime when the compiler determines that the string is potentially vulnerable to memory safety issues. This approach significantly reduces the overhead of ASan while maintaining its effectiveness in detecting memory safety bugs.

While the intended execution path should not incur any overhead due to compiler optimizations, a few local variables solely accessed by ASan helper functions are not currently removed and we rely on optimizing them away. A future commit may reduce number of these variables.

Commit:
[ASan][libc++] Update string ASan annotations to Zero-overhead

This commit introduces Zero-overhead string ASan annotations by wrapping every call to ASan helper functions inside a constexpr if statement. This ensures that ASan checks are only performed at runtime when the compiler determines that the string is potentially vulnerable to memory safety issues. This approach significantly reduces the overhead of ASan while maintaining its effectiveness in detecting memory safety bugs.

While the intended execution path should not incur any overhead due to compiler optimizations, a few local variables solely accessed by ASan helper functions are not currently removed and we rely on optimizing them away. A future commit may reduce number of these variables.

EricWF commented on libcxx/include/string:
```diff
@@ -659,6 +659,13 @@ _LIBCPP_PUSH_MACROS
 #endif
+#ifndef _IF_ASAN
+#  ifndef _LIBCPP_HAS_NO_ASAN
+#    define _IF_ASAN() if _LIBCPP_CONSTEXPR (true)
```
Please put this in __config. And you were right, we should name this `_LIBCPP_IF_ASAN()`.

AdvenamTacet replied:
Moved back to string file based on: https://github.com/llvm/llvm-project/pull/76165#discussion_r1434348292

EricWF commented on libcxx/include/string:
```diff
@@ -858,7 +865,7 @@ private:
       __set_long_size(__size);
     }
-    __annotate_new(__size);
+    _IF_ASAN() __annotate_new(__size);
```
Please always use `{ ... }` to scope these. The macro introduces enough magic.

EricWF commented on libcxx/include/string:
```diff
@@ -936,12 +949,12 @@ public:
-      __str.__annotate_new(0);
-      if (!__is_long() && this != &__str)
+      _IF_ASAN() __str.__annotate_new(0);
+      _IF_ASAN() if (!__is_long() && this != &__str)
```
Just use a single block here and wrap it in ` { ... }`.

AdvenamTacet replied:
Done.

EricWF reviewed (changes requested):
I think this works great. I don't love the way it looks but we can iterate on it.

The changes I would like before landing are:

(1) Rename to `_LIBCPP_IF_ASAN`.
(2) Put it in __config.
(3) Always use `{ ... }` after `_LIBCPP_IF_ASAN()`.

I think (3) is important because it's not clear exactly what `_LIBCPP_IF_ASAN()` expands to. We don't know if it actually captures the next statement as part of an `if` or not.

EricWF commented:
Also we need to do this to *everything*. I think everything needs to compile away fully without the optimizer before I feel comfortable fixing forward with this.

EricWF commented:
I'm putting up a change shortly to disable `-Wc++17-extensions`.

Commit:
Fixes based on the code review

(1) Renamed to `_LIBCPP_IF_ASAN`.
(2) Moved to `__config`.
(3) Added `{ ... }` after `_LIBCPP_IF_ASAN()`.

Missing: `#ifdef` around ASan related variables.

Commit:
Conditionalize ASan variables

Variables used only by ASan helper functions are conditionalized (put behind `#ifndef _LIBCPP_HAS_NO_ASAN`) after that commit.
That should remove any impact on performance.

ldionne commented on libcxx/include/__config:
```diff
@@ -1490,6 +1490,16 @@ __sanitizer_verify_double_ended_contiguous_container(const void*, const void*, c
 #  endif
 
+// ASan related macros
+// _LIBCPP_IF_ASAN is used to removed ASan related code, whenever compiled without ASan.
```
Since this is not suuuuper clean, I would suggest keeping this inside `<string>` and name it something string-specific like `_LIBCPP_STRING_IF_ASAN`. Hopefully in the future we can just clean it up.

EricWF replied:
Sorry, I suggested it be moved. But lets go with Louis suggestion.

AdvenamTacet replied:
I don't have an opinion about file, but so long name makes it less readable and I'm not a fan of it. But if we keep it inside string file, I understand that may be the best case, so I just did that, just pointing it out.

EricWF commented on libcxx/include/__config:
```diff
@@ -1490,6 +1490,16 @@ __sanitizer_verify_double_ended_contiguous_container(const void*, const void*, c
 
+// ASan related macros
+// _LIBCPP_IF_ASAN is used to removed ASan related code, whenever compiled without ASan.
+#ifndef _LIBCPP_IF_ASAN
+#  ifndef _LIBCPP_HAS_NO_ASAN
+#    define _LIBCPP_IF_ASAN() if _LIBCPP_CONSTEXPR (true)
+#  else
+#    define _LIBCPP_IF_ASAN() if _LIBCPP_CONSTEXPR (false)
+#  endif
+#endif
```
We should gaurd this with `_LIBCPP_INSTRUCTMENTED_WITH_ASAN` too since the underlying calls to asan's runtime only occur if that's the case. So no point turning this on otherwise.

EricWF commented:
The tests should start passing for this once #76178 lands.

Commit:
Changes based on @ldionne code review

- Macro moved to from `__config` to `string` file.
- `_LIBCPP_IF_ASAN` -> `_LIBCPP_STRING_IF_ASAN`.
- No more check `#ifndef _LIBCPP_STRING_IF_ASAN` before defining macro.

Commit:
Remove additional newline

EricWF reviewed (commented):
Once the bots pass I'm good with this. That's contingent on #76178 being accepted.

Thank you again for working on this.

Commit:
if constexpr since C++17

AdvenamTacet commented on libcxx/include/string:
```diff
@@ -2075,27 +2095,43 @@ private:
 
   _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 basic_string&
   __null_terminate_at(value_type* __p, size_type __newsz) {
+#ifndef __LIBCPP_HAS_NO_ASAN
```
We definitely should consider how to refactor it. I can reduce number of lines like that by removing a few not necessary variables, in another patch soon, but not completely remove all cases.

ldionne commented:
I would prefer that we understand whether https://github.com/llvm/llvm-project/pull/76082 is sufficient first, and if it's not, then we can go ahead with this change. I'm worried about this change because it's really reducing the readability of our string implementation, which at this point is already borderline unmanageable. So I would only do this if we absolutely must, and that would need to include the justification for this change, i.e. what concrete problem we are solving.

AdvenamTacet commented:
We upstreamed https://github.com/llvm/llvm-project/pull/76082 and https://github.com/llvm/llvm-project/pull/76200 - @EricWF is there still a need for that PR or can we close it without merge?

EricWF commented:
We can close it without merging. Thanks for your work on this.
